### PR TITLE
InterestAccrual: use more flexible interest rate type

### PR DESCRIPTION
**Pull request by offerijns**
# Description

Address https://github.com/centrifuge/centrifuge-chain/issues/1097

### CODE REVIEW

File: libs/traits/src/lib.rs
Comment (lemunozm) [inline libs/traits/src/lib.rs:333]: I think could be great to add a `map/try_map` function to `InterestRate` and let the caller call the operation they want over the rate. Similar to https://github.com/centrifuge/centrifuge-chain/pull/1435/files#diff-14aad6e50ab4015efc0d5939359c1676112c04751b9a825c15e1a588677da0b6R42

File: libs/traits/src/lib.rs
Comment (lemunozm) [inline libs/traits/src/lib.rs:384]: I think your previous question come because the name is not updated to `interest_rate_per_year` here.

File: libs/traits/src/lib.rs
Comment (lemunozm) [inline libs/traits/src/lib.rs:336]: Definitely, I would move now this trait and the `InterestRate` related types to their own module `cfg_traits::interest_rate`.

**Review by lemunozm — COMMENTED** [review]
Very good and clean and readable change, as a surgeon! 🙌🏻

Left a couple of minor changes that could be done in the future too.

I've seen the PR is over the other PR of principal & interest separation. Could we merge this in main once the other is merged? Just to not merge a lot of concepts in one PR.

File: pallets/interest-accrual/src/lib.rs
Comment (branan) [inline pallets/interest-accrual/src/lib.rs:439]: This method was already pretty deeply nested and it's worse now. Might be worth trying to break out the guts of it into its own function that just mutates the vector or something?

File: pallets/interest-accrual/src/lib.rs
Comment (branan) [inline pallets/interest-accrual/src/lib.rs:473]: As above

**Comment by offerijns** [comment]
> I've seen the PR is over the other PR of principal & interest separation. Could we merge this in main once the other is merged? Just to not merge a lot of concepts in one PR.

Yeah I think that's fine! Just based it on that PR because it contained relevant changes, but we can just rebase this one after merging that one.

File: pallets/interest-accrual/src/lib.rs
Comment (offerijns) [reply pallets/interest-accrual/src/lib.rs:439]: I guess it makes sense, but let's note that as a future clean up? Doesn't seem relevant particularly to this PR and this one is somewhat urgent.

File: libs/traits/src/lib.rs
Comment (offerijns) [reply libs/traits/src/lib.rs:333]: I actually prefer ensure add impl, because this is slightly more specific. The advantage is, say at some point there is also a floating rate variant next to the fixed variant. Then with map, it's unclear what field this should touch, and the implementor would need to handle both cases. Versus with ensure add the impl on interest rate can decide that adding always means increasing the inner rate.

Comment (offerijns) [reply libs/traits/src/lib.rs:333]: But we can still definitely also add `map/try_map` in the future, it can't hurt

File: libs/traits/src/lib.rs
Comment (mustermeiszer) [inline libs/traits/src/lib.rs:311]: Any reason why we are not implementeing the trait itself on the Type?

File: pallets/interest-accrual/src/lib.rs
Comment (mustermeiszer) [inline pallets/interest-accrual/src/lib.rs:516]: Rather for @branan and @lemunozm. Why are we using saturating math here?

File: libs/traits/src/lib.rs
Comment (lemunozm) [reply libs/traits/src/lib.rs:311]: You "cannot", this trait is implemented automatically for each T that already is `CheckedAdd`. You could implement `CheckedAdd` for this type, having automatically an `EnsureAdd` impl.

File: pallets/interest-accrual/src/lib.rs
Comment (lemunozm) [reply pallets/interest-accrual/src/lib.rs:516]: Because `MAX_INTEREST_RATE` is a const, we can ensure it's ok at compile time. But we could use `check_from_integer()` too, to be sure the constant value is not changed after the conversion

File: libs/traits/src/lib.rs
Comment (lemunozm) [reply libs/traits/src/lib.rs:341]: `map/try_map` usually only acts over the most important property. i.e. `map` only maps the ok branch in a `Result` type. For this type, the core field is the inner rate, so I think it makes sense to only map over it. Other properties are secondary. 

Anyways, it can makes sense to have an `ensure_add()` method directly in `InterestRate`. In that case, it can be built using `try_map` to factorize most of the code as follows:

```rust
fn ensure_add(&self, rate: Rate) -> Result<Self, DispatchError> {
    self.try_map(|rate_per_year| rate_per_year.ensure_add(rate))
}

fn ensure_sub(&self, rate: Rate) -> Result<Self, DispatchError> {
    self.try_map(|rate_per_year| rate_per_year.ensure_sub(rate))
}
```

Comment (lemunozm) [reply libs/traits/src/lib.rs:341]: Of course, any of this is required for this PR, and could be added in the future 👍🏻

File: libs/traits/src/lib.rs
Comment (offerijns) [reply libs/traits/src/lib.rs:295]: I'm sure you're right, but for my understanding, why?

Comment (lemunozm) [reply libs/traits/src/lib.rs:295]: For small types, such as simple enums, numbers, or pointers, making them `Copy` makes it easy to handle them later i.e. in iterators and vectors. As a rule of thumb, anything under 128bits can (should?) be copied. Things bigger can get worse performance because of the bytes copied, and a reference is preferred. For those cases is better to not add `Copy` and let the caller pass it as a reference or explicitly clone if needed.

An important detail is that Copy only can be added if all internal types are already `Copy` before. So, adding a non-copiable field to a `Copy` struct can lead to a big refactor of the code base.

Given said this, I would add `Copy` to `CompoundingSchedule` because it's like a simple mark that adds an information unit. But not sure about adding it to `InterestRate`.

File: libs/traits/src/lib.rs
Comment (mustermeiszer) [reply libs/traits/src/lib.rs:311]: Ah, right. Okay. Well not mandatory, but nice to have `CheckedAdd` then ^^

Comment (lemunozm) [reply libs/traits/src/lib.rs:311]: Although I would like to do it. It's is required to be `PartialOrd` + `CheckedAdd` to be `EnsureAdd`, and it's difficult to determine a `PartialOrd` with arithmetic meaning for a type with several fields.

This is something that now, after a lot of usage, I would design differently. Instead of making this trait automatic for any `T`, I would implement it for basic & substrate types, allowing then any custom implementation for more complex custom time.

Maybe I do a new PR to Substrate this change. I think it can be beneficial, although it will be a huge breaking change, and I do not know if they would approve it.

File: pallets/loans/src/entities/pricing/external.rs
Comment (lemunozm) [inline pallets/loans/src/entities/pricing/external.rs:53]: Tech debt from #1450 here

**Review by mustermeiszer — COMMENTED** [review]
> From the issue, Jeroen says that we should also change the rate from pool's Tranches. Nevertheless, changing things in the pool will require a migration (or maybe not?), and by now we were not touching other tnings from the pool, and could obstaculize more the release. I think we could leave this for a future pool's PR although it will lead to a migration.

Completely agree, that we should not touch pools right now. Although, this will make migrating later on hellish, but I guess we can not prevent that. Refactoring the pool will be a big future task

Regarding the rest. 

* I do not really like using the `enum InterestRate` as a fixed type in the pallet and trait interface. Does not feel like the right change.


Wdyt about creating a a trait that bounds the parameter to satisfy all needs. Something like

```rust
trait Rate {
     type Interest: FixedPointNumberExt;

     fn steps_per_year() -> impl FixedPointOperand;

     fn interest_rate_per_year(&self) -> Self::Interest;

     fn steps(&self, last: Moment, now: Moment) -> usize;
}
```

So `validate` becomes:

```rust
pub(crate) fn validate_interest_rate(
			rate: T::Rate,
		) -> DispatchResult {
			let max_decimals = T::InterestRate::saturating_from_integer(T::MaxDecimals::get());
			let maximum = T::InterestRate::saturating_from_integer(MAX_INTEREST_RATE);
			ensure!(
				rate.interest_rate_per_year() <= maximum
					&& rate.interest_rate_per_year() >= Zero::zero()
					&&(rate.interest_rate_per_year().saturating_mul(max_decimals)).frac()
						== Zero::zero(),
				Error::<T>::InvalidRate
			);
			Ok(())
		}
```
Furthermore, we would need to adapt
* finding `Rates` based on steps per year and interest per step
* `on_initialize` calculation
* other ?

Is just my first thought, but we should be able to generailze the accrual pallet like that and we are not forced to do a migration once we want montly accrual and so on.

File: pallets/interest-accrual/src/lib.rs
Comment (lemunozm) [reply pallets/interest-accrual/src/lib.rs:168]: Actually, we do not need it. We can safely remove it.

**Comment by lemunozm** [comment]
Really like your idea of extracting the behavior of an interest rate into a trait!

Only two things:
- Does it make sense another implementation for the `Rate` trait, rather than the own `InterestRate`, right now I'm not able to think of another possible future implementation. If only one exists, not sure if generalizing helps here (although it's cool a lot). As a note, the idea behind the enum in `InterestRate` is that you can combine `Fixed` & `Variable` (or others) at the same time. So any implementation of `Rate` trait would need that "enum" or similar.
- In case of doing it, because it would not affect to the storage, could we leave it for a future PR after the next release?

**Comment by lemunozm** [comment]
> we are not forced to do a migration once we want montly accrual and so on.

Why are we not forced to do it with the trait?

**Comment by mustermeiszer** [comment]
> Why are we not forced to do it with the trait?

As we would also adapt right now what we store in storage under `RateDetails`. If not, then must migrate of course.

**Comment by mustermeiszer** [comment]
> Does it make sense another implementation for the Rate trait, rather than the own InterestRate, right now I'm not able to think of another possible future implementation. If only one exists, not sure if generalizing helps here (although it's cool a lot). As a note, the idea behind the enum in InterestRate is that you can combine Fixed & Variable (or others) at the same time. So any implementation of Rate trait would need that "enum" or similar.

Yes, every impl would need that. I was more thinking of already preparing interest accrual for `foss3`. But again it is not mandatory. It just felt like using an "incomplete" enum for the interface was a fix rather than a solution. 

I think adapting this to a trait will require a migration as we can no longer rely on `interest_rate_per_sec`. The same goes for the current change. 

As soon as we want another compounding schedule, we will need to migrate to something that reflects that rate and is able to determine how many compounding steps happened between `last_updated` and `now`

**Comment by lemunozm** [comment]
> It just felt like using an "incomplete" enum for the interface was a fix rather than a solution.

Yes I the idea from Jeroen was to prepare this for future work, adding this now before because of the release.

> As soon as we want another compounding schedule, we will need to migrate to something that reflects that rate and is able to determine how many compounding steps happened between `last_updated` and `now`

Although it would require migration in the `pallet-interest-accrual`, having this type yet in `pallet-loans` would avoid migrating it later there if we already have that "enum" in the type

**Comment by lemunozm** [comment]
Agree then in merging this now?

**Review by mustermeiszer — APPROVED** [review]

**Pull request merged**
